File: Uno/UnoSorter.java
import java.util.*; //Import
public class UnoSorter {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of cards: ");
        int n = sc.nextInt();
        sc.nextLine(); // Consume newline
        
        List<String> cards = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            System.out.print("Enter card " + (i+1) + " (e.g., 'red 3', 'wild draw 4'): ");
            cards.add(sc.nextLine().trim().toLowerCase());
        }
        
        // Sort: first by color total points asc, then stable by individual value asc
        cards.sort((a, b) -> {
            String colorA = getColor(a);
            String colorB = getColor(b);
            int totalA = getColorTotal(cards, colorA);
            int totalB = getColorTotal(cards, colorB);
            if (totalA != totalB) return Integer.compare(totalA, totalB);
            return Integer.compare(getValue(a), getValue(b));
        });
        
        System.out.println("\nSorted hand (lowest color total points first, then by value):");
        for (String card : cards)  System.out.println(card);
        sc.close();
    }
    
    static int getValue(String card) {
        if (card.contains("draw 4") || card.contains("wild draw 4")) return 50;
        if (card.contains("+2") || card.contains("draw 2") || card.contains("skip") || card.contains("reverse") || card.contains("wild")) return 20;
        String[] parts = card.split(" ");
        if (parts.length >= 2 && parts[1].matches("\\d+")) return Integer.parseInt(parts[1]);
        return 0; 
    }
    
    static String getColor(String card) {
        if (card.contains("wild") || card.contains("draw 4")) return "Wild";
        String[] parts = card.split(" ");
        if (parts.length > 0) {
            String first = parts[0].substring(0, 1).toUpperCase();
            if ("RGBY".contains(first)) return first;
        }
        return "Wild";
    }
    
    static int getColorTotal(List<String> cards, String color) {
        return cards.stream()
                    .filter(c -> getColor(c).equals(color))
                    .mapToInt(UnoSorter::getValue)
                    .sum();
    }
}

File: Uno/UnoSorter.py
def get_value(card):
    card = card.lower().strip()
    if 'draw 4' in card or 'wild draw 4' in card:
        return 50
    if any(x in card for x in ['+2', 'draw 2', 'skip', 'reverse', 'wild']):
        return 20
    parts = card.split()
    if len(parts) >= 2 and parts[1].isdigit():
        return int(parts[1])
    return 0

def get_color(card):
    card = card.lower().strip()
    if 'wild' in card or 'draw 4' in card:
        return 'Wild'
    parts = card.split()
    if parts:
        first = parts[0][0].upper()
        if first in 'RGBY':
            return first
    return 'Wild'

cards = []
n = int(input("Enter number of cards: "))
for i in range(n):
    card = input(f"Enter card {i+1} (e.g., 'red 3', 'wild draw 4'): ")
    cards.append(card)

def color_total(color):
    return sum(get_value(c) for c in cards if get_color(c) == color)

# Sort by color total asc, then value asc (stable)
sorted_cards = sorted(cards, key=lambda c: (color_total(get_color(c)), get_value(c)))

print("\nSorted hand (lowest color total points first, then by value):")
for card in sorted_cards:
    print(card)
